File: leetcode/20.valid-parentheses.py
# @leet start
class Solution:
    def isValid(self, s: str) -> bool:
        BRACE_DICT = {")": "(", "}": "{", "]": "["}
        stack: list[str] = []

        for ch in s:
            if ch in BRACE_DICT:
                if not stack or stack[-1] != BRACE_DICT[ch]:
                    return False

                stack.pop()
            else:
                stack.append(ch)

        return len(stack) == 0


# @leet end

File: leetcode/20.valid-parentheses.ts
// @leet start
function isValid(s: string): boolean {
  type Open = "(" | "[" | "{";
  type Close = ")" | "]" | "}";
  const PAREN_MAP: Record<Open, Close> = {
    "(": ")",
    "[": "]",
    "{": "}",
  };
  const stack: Open[] = [];

  for (const char of s) {
    if (char in PAREN_MAP) {
      stack.push(char as Open);
    } else {
      const left = stack.pop();
      if (left === undefined || char !== PAREN_MAP[left]) {
        return false;
      }
    }
  }

  return stack.length === 0;
}
// @leet end
